lib/saluki-components/src/transforms/name_filter/mod.rs

```diff
@@ -0,0 +1,238 @@
+        Ok(Box::new(NameFilter {
+            metric_prefix: self.metric_prefix.clone(),
+            metric_prefix_blacklist: self.metric_prefix_blacklist.clone(),
+            blocklist: Blocklist::new(&self.metric_blocklist, self.metric_blocklist_match_prefix),
+        }))
+    }
+}
+
+impl MemoryBounds for NameFilterConfiguration {
+    fn specify_bounds(&self, builder: &mut MemoryBoundsBuilder) {
+        // Capture the size of the heap allocation when the component is built.
+        builder.minimum().with_single_value::<NameFilter>();
+    }
+}
+
+#[derive(Debug)]
+struct Blocklist {
+    data: Vec<String>,
+    match_prefix: bool,
+}
+
+impl Blocklist {
+    fn new(data: &[String], match_prefix: bool) -> Self {
+        let mut data = data.to_owned();
+        data.sort();
+
+        if match_prefix && !data.is_empty() {
+            let mut i = 0;
+            for j in 1..data.len() {
+                if !data[j].starts_with(&data[i]) {
+                    i += 1;
+                    data[i] = data[j].clone(); // Move item to next position
+                }
+            }
+            data.truncate(i + 1); // Only keep valid elements
+        }
+
+        Blocklist { data, match_prefix }
+    }
+
+    fn test(&self, name: String) -> bool {
+        if self.data.is_empty() {
+            return false;
+        }
+
+        let i = self.data.binary_search(&name);
+
+        // Check for prefix match when match_prefix is true
+        if self.match_prefix {
+            if let Ok(index) = i {
+                if index > 0 && name.starts_with(&self.data[index - 1]) {
+                    return true;
+                }
+            } else if i.is_err() {
+                let index = i.unwrap_err();
+                if index > 0 && name.starts_with(&self.data[index - 1]) {
+                    return true;
+                }
+            }
+        }
+
+        if let Ok(index) = i {
+            return name == self.data[index];
+        }
+
+        false
+    }
+}
+
+#[derive(Debug)]
+pub struct NameFilter {
+    metric_prefix: String,
+
+    metric_prefix_blacklist: Vec<String>,
+
+    blocklist: Blocklist,
+}
+
+impl NameFilter {
+    fn enrich_metric(&self, mut metric: Metric) -> Option<Metric> {
+        let metric_name = metric.context().name().deref();
+        let mut new_metric_name = metric_name.to_string();
+
+        if !self.is_excluded(metric_name) {
+            new_metric_name = format!("{}{}", self.metric_prefix, metric_name);
+        }
+
+        if self.blocklist.test(new_metric_name.clone()) {
+            debug!("Metric {} excluded due to blocklist.", new_metric_name);
+            return None;
+        }
+
+        // Update metric with new name.
+        let new_context = metric.context().with_name(new_metric_name);
+        let existing_context = metric.context_mut();
+        *existing_context = new_context;
+
+        Some(metric)
```

Comment:
What we really want to avoid doing here is allocating or recreating anything if we didn't actually modify anything. So we know that we're only going to modify the metric name if we have a prefix configured, so we should probably have that be our top-level conditional.

```rust
if self.metric_prefix.is_empty() {
  // just check if this metric is in the blocklist, return Some/None based on that
} else {
  // existing logic goes here
}
```

That gets us away from having to do gymnastics to generate `metric_name`/`new_metric_name` prior to passing it to `self.blocklist.test(..)`.

Reply:
You might need to tweak `is_excluded` a little bit.